### ISSUE TITLE
Use Gists for easy customization

### ISSUE THREAD

**Issue by poelstra** [FirstLegoLeague/displaySystem#2]
Just as a crazy idea, wouldn't it be cool to use e.g. a GitHub Gist (through e.g. http://bl.ocks.org) or jsFiddle to 'host' a custom version of the display system?

Similar to how http://chilipeppr.com/ provides hardware fiddles.

That way, people could start out with the demo system, then fork and edit as needed.

Especially if new display modules could be added by pointing to another Gist and/or Github repo!

I'm thinking that these external modules could provide an index.js that exports a function, and call that function with e.g. 'the displaySystem singleton' as its first argument to inject it.

Would be cool if one could go to e.g. http://firstlegoleague.github.io/displaySystem/#my-gist-id to load a certain configuration from there, and that configuration could point to a bunch of other gists/repos for more modules etc.

Trouble with e.g. ChiliPeppr is that it is apparently not trivial to make that stack work locally, which is a pretty hard requirement for our use-case, but I think we don't have that problem in our stack.

**Comment by rikkertkoppes** [comment]
Played with the idea. Direct gists don't  work since the raw links don't
provide a content-type. Haven't looked further into other options though.

This would be especially nice for themes, which I will work out next
(basically providing a custom style sheet)

**Comment by rikkertkoppes** [comment]
There is http://githack.com and https://rawgit.com/

https://gist.githubusercontent.com/rikkertkoppes/9c7e4efaba9dbbc4831b/raw/a50314f89a0e2a1d7d75e7cd748e63fc6eba8fa2/custom.css
becomes
https://rawgit.com/rikkertkoppes/9c7e4efaba9dbbc4831b/raw/a50314f89a0e2a1d7d75e7cd748e63fc6eba8fa2/custom.css
and (on a cdn)
https://cdn.rawgit.com/rikkertkoppes/9c7e4efaba9dbbc4831b/raw/a50314f89a0e2a1d7d75e7cd748e63fc6eba8fa2/custom.css

or (via githack)
https://gist.githack.com/rikkertkoppes/9c7e4efaba9dbbc4831b/raw/a50314f89a0e2a1d7d75e7cd748e63fc6eba8fa2/custom.css

However, all solutions urge to not use it in production. Also, I don't have a good idea how to set this up in an easy-to-use manner

**Comment by rikkertkoppes** [comment]
Another option would be to just use the json api. This works because cors headers are correctly set for the api. Just get the file contents, and insert as style element. No need to host it.

**Comment by rikkertkoppes** [comment]
Implemented, 83c6140d5ea095cb845ed07353cfff304ce7d342

**Comment by rikkertkoppes** [comment]
Reopened. The config in gist idea is still pretty awesome. Maybe pull this feature upwards

**Comment by rikkertkoppes** [comment]
Idea : create a generic scripts and style sheets loader that can either load as data url from a cache or from real url.

The gist can be used to populate that cache 

@poelstra what do you think?

**Comment by poelstra** [comment]
@rikkertkoppes I'm not exactly sure what you mean with the cache, but yes, the possibility of specifying one or more urls to load modules from (functionality and/or styles) was the intention of this issue.

**Comment by poelstra** [comment]
To clarify, let me expand on the use case I had in mind.

For example, for a typical FLL Benelux final, we have one display showing:
- a twitter feed at the bottom of the screen
- a countdown timer at the top-right corner
- the current wall-clock time at the top-left corner
- a message bar that may pop up at the bottom
- optionally a large scores display in the middle of the screen
- optionally a sponsors/volunteers-thank-you-image in the middle of the screen
- transparent background (currently a color-key) for mixing onto live video

We also have one or more displays in pit areas, where we also have some of these, but in that case we mostly don't have the transparent background, and instead constantly display the scores.
And/or, we have a screen for the jury area, that shows the time schedule for presentations, and also the twitter feed, but not the countdown timer.

So, my idea would be to have one Gist (or some other easily editable source of configuration) for each of these types of displays, containing a list of modules to load, and their configuration.
Modules can be e.g. the twitter feed, clock, timer, etc. Their configuration can be things like id (for control via mhub), position, size, source-url (e.g. for fetching tweets), etc.

Maybe it's possible to have themes as a type of module too, that determine e.g. main colors, fonts, etc. I can imagine, though, that theming can be too functional-block-specific, so maybe the Gist should always 'point' to an already-themed-instance of a piece of functionality (e.g. to a scores-display-block that is already styled in some way), or theming info could be part of the configuration data.

So, the idea is that a more tech-savy person can prepare e.g. a Gist (or multiple), with a certain combination of modules for a specific type of event. Other people can then fork the gist, edit some smaller things (add/remove a block, change a color or URL, change the URL of the local mhub instance, etc.) and then set that as the 'source' of their display(s), e.g. by passing it in the URL like http://firstlegoleague.github.io/displaySystem/#my-gist-id

**Comment by poelstra** [comment]
@rikkertkoppes Hmmm, about that cache, you mean to have the content available offline too, in case the Internet link fails during the day? (Which it typically does :))

**Comment by rikkertkoppes** [comment]
No, meant more like a dictionary. If the module name (which expands to a
filename) is in the gist resource, use that one via data url. Otherwise,
load via current route (real url with basepath)

For offline stuff, we should use localstorage to store gists indeed. Or
discourage gists and prefer people to copy them locally.

**Comment by rikkertkoppes** [comment]
Also, I'd like themes to be a two layer thing. First would be the theme
itself, the second a customization stylesheet to e.g. change only the color
scheme.

**Comment by poelstra** [comment]
> No, meant more like a dictionary. If the module name (which expands to a
> filename) is in the gist resource, use that one via data url. Otherwise,
> load via current route (real url with basepath)

I still don't know what you mean. Can you give an example?

**Comment by rikkertkoppes** [comment]
@poelstra after my explanation yesterday, how would I better rephrase this?
